src/devices/src/virtio/net/iovec.rs

```diff
@@ -0,0 +1,225 @@
+use std::ops::Deref;
+use vm_memory::{GuestMemory, GuestMemoryError, GuestMemoryMmap};
+
+#[derive(Debug)]
+pub enum Error {
+    /// We found a write-only descriptor where read-only was expected
+    WriteOnlyDescriptor,
+    /// An error happened with guest memory handling
+    GuestMemory(vm_memory::GuestMemoryError),
+}
+
+impl From<GuestMemoryError> for Error {
+    fn from(err: GuestMemoryError) -> Self {
+        Error::GuestMemory(err)
+    }
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+/// This is essentially a wrapper of a `Vec<IoSlice>` which can be passed `writev`.
+///
+/// It describes a buffer passed to us by the guest that is scattered across multiple
+/// memory regions. Additionally, this wrapper provides methods that allow reading arbitrary ranges
+/// of data from that buffer.
+#[derive(Debug)]
+pub(crate) struct IoVecBuffer<'a> {
+    // container of the memory regions included in this IO vector
+    vecs: Vec<IoSlice<'a>>,
+    // Total length of the IoVecBuffer
+    len: usize,
+}
+
+impl<'a> Deref for IoVecBuffer<'a> {
+    type Target = [IoSlice<'a>];
+
+    fn deref(&self) -> &Self::Target {
+        self.vecs.as_slice()
+    }
+}
+
+#[cfg(test)]
+impl<'a> From<&'a [u8]> for IoVecBuffer<'a> {
+    fn from(buf: &'a [u8]) -> Self {
+        Self {
+            vecs: vec![IoSlice::new(buf)],
+            len: buf.len(),
+        }
+    }
+}
+
+impl<'a> IoVecBuffer<'a> {
+    /// Create an `IoVecBuffer` from a `DescriptorChain`
+    pub fn from_descriptor_chain(mem: &'a GuestMemoryMmap, head: DescriptorChain) -> Result<Self> {
+        let mut vecs = vec![];
+        let mut len = 0usize;
+
+        let mut next_descriptor = Some(head);
+        while let Some(desc) = next_descriptor {
+            if desc.is_write_only() {
+                return Err(Error::WriteOnlyDescriptor);
+            }
+
+            // This is safe since we get `ptr` from `get_slice` which also checks the length
+            // of the descriptor
+            let ptr = mem.get_slice(desc.addr, desc.len as usize)?.as_ptr();
+            let slice = unsafe { std::slice::from_raw_parts(ptr, desc.len as usize) };
+            vecs.push(IoSlice::new(slice));
+            len += desc.len as usize;
+
+            next_descriptor = desc.next_descriptor();
+        }
+
+        Ok(Self { vecs, len })
+    }
+
+    /// Get the total length of the memory regions covered by this `IoVecBuffer`
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Reads a number of bytes from the `IoVecBuffer` starting at a given offset.
+    ///
+    /// This will try to fill `buf` reading bytes from the `IoVecBuffer` starting from
+    /// the given offset.
+    ///
+    /// # Retruns
+    ///
+    /// The number of bytes read (if any)
+    pub fn read_at(&self, buf: &mut [u8], offset: u64) -> Option<usize> {
+        // We can't read past the end of this `IoVecBuffer`
+        if offset >= self.len() as u64 {
+            return None;
```

Comment:
Also possibly a Rust idiom/nuance, this would have struck me as an error use-case. Do we want to subtly return nothing here? If so, maybe just specify a little in the doc comment.

Reply:
you are right, in that this functionality in other languages would return an Error, probably because of lack of the `Option` type. In this case not finding enough space might not necessarily be an Error. _We ran out of space, so try again later_

Reply:
Not a strong preference on my end.